### ISSUE TITLE
MetaMask says not connected when the instance which has connected it is in a iframe

### ISSUE THREAD

**Issue by joshstevens19** [MetaMask/metamask-extension#10942]
Hey! 

If you call  `eth_requestAccounts` within an iframe of the website MetaMask connects, lets you sign, etc, and all the functionality works but then will say not connected on the browser extension and never bring up the UI to reconnect if you switch accounts. It thinks the parent domain is the thing that requested access to it which is incorrect it's actually an iframe within the parent website which has access to it. 

![image](https://user-images.githubusercontent.com/18472422/116407656-e4c25a80-a829-11eb-9c6a-c41a8894241e.png)

Our wallet is integrated using iframes and we want to integrate the ability to use MM on it. But without this logic, we can not go any further as the UX/UI will just end up being impossible to maintain (due to not being able to reconnect or your UI showing) 

A simple way to recreate:

- create a simple app which is the main page
- create another app that on page load calls `await window.ethereum.request({ method: 'eth_requestAccounts' })`
- now look at your connected sites for that account and try to change accounts to toggle the reconnect logic

I personally think the solution should be making sure you look at all iframes embedded into the website as well as the main domain this way you can tell if connected site x is actually in the website meaning you can allow it to work within iframes.

**Comment by danfinlay** [comment]
Interesting. We probably could iterate recursively through iframes to check for connected domains, but then what if there are two connected domains? Sounds like there could be a lot of complexity unraveled  in doing what you suggest.

Could you give me a better sense of why this is necessary?

One workaround idea: Have the main domain request connection, and then forward its messages to your iframe as a form of delegation. (Also return its messages).

**Comment by joshstevens19** [comment]
So I was thinking if the main domain is connected then it stops looking, but then it loops through all iframes, and the first one it finds which matches a connected domain it then stats that as the domain. Right now even though our iframe is connected to say account 1 MM thinks the parent domain is the only thing that has the ability to connect to it which is not true. This use to work before your permission approval account modal came in. 

So our wallet is in an iframe, the main domain can not forward certain requests like for example if you click in our wallet to send a transaction that action is in the wallet itself and not the dApp, you can see here how we can not do that workaround. Also, we are not in control of the dApp that integrate with our wallet (its open-source, and anyone can come up and setup a dApp), it's up to them to write logic so we would not want to force them to write bespoke MM logic for this to work with our wallet. 

I think I'm basically asking for iframes to be able to work connected to MM which right now does work but the connection logic is incorrect. If it has approved iframe1.com but is on main-domain.com I would expect it to be able to work out that main-domain.com has no connection and iframe1.com is within the context here so is connected. 

If you always priorities the main domain THEN look for iframes and stop when you find the first connected one I don't see any super complexity unraveling! 

Let me know your thoughts!

**Comment by ihorbond** [comment]
hey @joshstevens19 what did you end up doing ? we have similar problem. We provide an NFT minting page to our customers as an iframe they can drop into any site builder like wix, webflow, etc. They won't know how to implement workarounds so it would be great if Metamask natively supported this @danfinlay

**Comment by joshstevens19** [comment]
Hey @ihorbond i ended up leaving the company this feature was for but I think due to this we didn’t support metamask. I think the issue is still a problem!

**Comment by ihorbond** [comment]
@joshstevens19 yeah I'd leave them over MM connectivity issues as well! Hahaha just joking :) Curious: have you looked into posting messages between the main site and iframe site windows ? It gives me CORS errors since they are different domains but was hoping maybe if I set CORS to * it might actually work ?
